### ISSUE TITLE
Version 2.1 is backwards-incompatible

### ISSUE THREAD

**Issue by jsakars** [magento/magento2#5617]
[Versioning policy overview](http://devdocs.magento.com/guides/v2.0/architecture/versioning.html) clearly states that `"Magento software versioning complies with Semantic Versioning 2.0.0"`.

Recently version 2.1.0 was released. Based on [Semantic Versioning 2.0.0](http://semver.org/) it was supposed to be `"version when you add functionality in a backwards-compatible manner"`.

To customize default Magento behavior You would normally extend specific module, class etc. Take CMS page module for example, if You would have extended "save" action in version 2.0.7 and now would like to upgrade to version 2.1.0 (which by definition must be backwards compatible) You would actually get a fatal error: `PHP Fatal error:  Uncaught TypeError: Argument 3 passed to Magento\Cms\Controller\Adminhtml\Page\Save::__construct() must be an instance of Magento\Framework\App\Request\DataPersistorInterface, none given`.
- https://github.com/magento/magento2/blob/2.0/app/code/Magento/Cms/Controller/Adminhtml/Page/Save.php#L22
- https://github.com/magento/magento2/blob/2.1/app/code/Magento/Cms/Controller/Adminhtml/Page/Save.php#L38-L42

This is just one of the examples. Besides backend modules we see that our custom front-end theme (which works fine with 2.0.7) now gives fatal error.

Now to actually upgrade to version 2.1.0 You need to fix all backwards-incompatible overrides You have in Your custom modules and themes.

All this seems very off compared to Semantic Versioning used in modern 3rd party PHP libraries.

It seems that in core this is related to the fact that bugs found in certain version e.g. 2.0.x are fixed in mainstream `develop` branch which also holds new features and backwards-incompatibilities - releases are then cut from `develop` branch and here we are...

**Comment by hostep** [comment]
@werdlv: semver applies to the module versions, not to the big Magento versions which are being used for marketing purposes (imagine that for this change they had to release Magento 3.0, would be a bit silly)

Anyway, they applied semver correctly in this case, see:
[magento/module-cms in Magento 2.0.x](https://github.com/magento/magento2/blob/2.0/app/code/Magento/Cms/composer.json#L21) vs [magento/module-cms in Magento 2.1.x](https://github.com/magento/magento2/blob/2.1/app/code/Magento/Cms/composer.json#L21)

**Comment by jsakars** [comment]
@hostep as a technical user (developer) You don't care about version for each specific Magento module as You don't depend on them directly - Magento CE does. Normally user is only concerned with `"magento/product-community-edition"` dependency.

Also it's seems very strange that [Upgrade to Magento version 2.1](http://devdocs.magento.com/guides/v2.0/release-notes/tech_bull_21-upgrade.html) (which clearly isn't a post for marketing) doesn't say a word regarding backward incompatible changes.

**Comment by adragus-inviqa** [comment]
I get your frustration. Had it, too. But...

They cannot maintain BC. Simple. Not at this stage. There's too much to update (M1 cruft). They're trying _really_ hard from what I can tell, but they can't always keep BC _and_ modernise M2 codebase, both at the same time.

Even if they stood by semver (which they don't), semver itself is not a guarantee, merely a _pursuit_.

PS: I could be completely wrong on the above. I'm not a core dev.

**Comment by hostep** [comment]
@werdlv: sorry I wholeheartedly have to disagree here. As a developer, you actually _should_ care about specific modules and their versions, since your custom code is most likely depending on one or multiple of those modules. Your custom module is built for a specific version of those modules. So whenever one of these Magento modules gets a backwards-incompatible update, you are forced to review your custom module and will have to test it to see if it still works and fix it if necessary and then also update the version you depend on in your modules `composer.json` file.
So, in your custom module, you should definitely define dependencies on those specific Magento modules with the specific version you tested it on and certainly _not_ on the Magento product version itself, that's definitely not how you should work.

I would give it some time, try it out, play with it, people will have to grow accustomed to how this is going to work and a lot of developers will have difficulties with this. But in the end, it should all make sense.

Now, mistakes will probably been made in the Magento 2 modules to keep to the semver idea. And also in the custom modules people develop. But since Magento 2 is still so new, this is understandable, and those mistakes will happen less and less in the future I hope.

You can even see this now, a Magento 2 fixed version can exist in multiple different variants, since it itself depends on 3rd party modules, for example a couple of symfony modules which it uses.
If you have set up Magento 2.1.0 a couple of weeks ago, and then now ask composer to look for updates of all the dependencies, it actually will find a bunch of modules with backwards-compatible updates from symfony for example, even though nothing has changed in the Magento 2 code itself.
That's one of the beauties of the current composer dependency model: we don't have to wait on the Magento team to get bugfixes in place for 3rd party modules, we can now take care of this ourselves as developers. But this does put more responsibility in our hands of course.

The Magento product version is only a marketing thing. As a developer, you really shouldn't care about that version.

**Comment by orlangur** [comment]
@werdlv you didn't read the http://devdocs.magento.com/guides/v2.0/architecture/versioning.html carefully. Only methods marked with `@api` are guaranteed to remain compatible. You customize anything besides this at your own risk.

Also, https://github.com/magento/magento2/issues/4228#issuecomment-232931191 may give you some insight on what is going on with BC policy/painful upgrades.

**Comment by davidalger** [comment]
The GitHub issue tracker is intended for technical issues only. Please refer to the [Community Forums](https://community.magento.com) or [Magento Stack Exchange](http://magento.stackexchange.com/) site for technical questions and discussion. Closing as off-topic.
